x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java

```diff
@@ -187,22 +178,25 @@ protected Query doToQuery(SearchExecutionContext context) throws IOException {
         } else {
             return organicQuery.toQuery(context);
         }
-
     }
 
-    @SuppressWarnings("unchecked")
     @Override
-    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
-        if (pinnedIds != null || pinnedDocs != null) {
-            return this;
-        } else if (pinnedIdsSupplier != null || pinnedDocsSupplier != null) {
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        if (pinnedIdsSupplier != null || pinnedDocsSupplier != null) {
             List<String> identifiedPinnedIds = pinnedIdsSupplier != null ? pinnedIdsSupplier.get() : null;
             List<Item> identifiedPinnedDocs = pinnedDocsSupplier != null ? pinnedDocsSupplier.get() : null;
             if (identifiedPinnedIds == null && identifiedPinnedDocs == null) {
-                return this; // not executed yet
-            } else {
-                return new RuleQueryBuilder(organicQuery, matchCriteria, rulesetId, identifiedPinnedIds, identifiedPinnedDocs, null, null);
-            }
+                return this; // Not executed yet
+            } else if ((identifiedPinnedIds != null && identifiedPinnedIds.isEmpty())
+                && (identifiedPinnedDocs != null && identifiedPinnedDocs.isEmpty())) {
+                    return organicQuery; // Nothing to pin here
```

Comment:
We should check
```
 if ((pinnedIds != null && pinnedIds.isEmpty() == false) && (pinnedDocs != null && pinnedDocs.isEmpty() == false)) {
            throw new IllegalArgumentException("applied rules contain both pinned ids and pinned docs, only one of ids or docs is allowed");
}
```

As that will never be checked and we arbitrarily (seemingly) `identifiedPinnedIds` first if both are supplied.